modal/functions.py

```diff
@@ -1660,7 +1733,57 @@ def wrapper(raw_f: Callable[..., Any]) -> _PartialFunction:
     return wrapper
 
 
+@typechecked
+def _build(
+    _warn_parentheses_missing=None,
+) -> Callable[[Union[Callable[[Any], Any], _PartialFunction]], _PartialFunction]:
+    if _warn_parentheses_missing:
+        raise InvalidError("Positional arguments are not allowed. Did you forget parentheses? Suggestion: `@build()`.")
+
+    def wrapper(f: Union[Callable[[Any], Any], _PartialFunction]) -> _PartialFunction:
+        if isinstance(f, _PartialFunction):
+            return f.add_flags(_PartialFunctionFlags.BUILD)
+        else:
+            return _PartialFunction(f, _PartialFunctionFlags.BUILD)
+
+    return wrapper
+
+
+@typechecked
+def _enter(
+    _warn_parentheses_missing=None,
+) -> Callable[[Union[Callable[[Any], Any], _PartialFunction]], _PartialFunction]:
+    if _warn_parentheses_missing:
+        raise InvalidError("Positional arguments are not allowed. Did you forget parentheses? Suggestion: `@enter()`.")
+
+    def wrapper(f: Union[Callable[[Any], Any], _PartialFunction]) -> _PartialFunction:
+        if isinstance(f, _PartialFunction):
+            return f.add_flags(_PartialFunctionFlags.ENTER)
+        else:
+            return _PartialFunction(f, _PartialFunctionFlags.ENTER)
+
+    return wrapper
+
+
+# TODO(erikbern): last argument should be Optional[TracebackType]
+ExitHandlerType = Callable[[Any, Optional[Type[BaseException]], Optional[BaseException], Any], None]
+
+
+@typechecked
+def _exit(_warn_parentheses_missing=None) -> Callable[[ExitHandlerType], _PartialFunction]:
+    if _warn_parentheses_missing:
+        raise InvalidError("Positional arguments are not allowed. Did you forget parentheses? Suggestion: `@exit()`.")
+
+    def wrapper(f: ExitHandlerType) -> _PartialFunction:
+        return _PartialFunction(f, _PartialFunctionFlags.EXIT)
+
+    return wrapper
+
+
 method = synchronize_api(_method)
 web_endpoint = synchronize_api(_web_endpoint)
 asgi_app = synchronize_api(_asgi_app)
 wsgi_app = synchronize_api(_wsgi_app)
+build = synchronize_api(_build)
```

Comment:
Why are these synchronized btw?

Reply:
I had a similar question actually. I think it has something to do with how we want `_PartialFunction.__get__` to return a `Function` not a `_Function`, which means we have to synchronize the `_PartialFunction` class, which means we end up with internal and external versions of everything... I'm not entirely sure though.

This is obviously pretty complex and it would be nice to simplify it